BROOKLYN-379: entity type includes OSGi prefix

aledsage opened the pull request.

bostko committed:
Clean up duplicated code in JcloudsLocationCustomizers

grkvlt committed:
Added configuration for template substitution variable map

ahgittin committed:
minor tidies to utils

ahgittin committed:
add static Maybe.getException(maybe) to avoid messy casts to Absent, and Asserts methods

aledsage committed:
Subscription callbacks: task has contextEntity tag

neykov committed:
Test executing a task in a context different from its own

When executing a task which is already tagged with a context entity in the execution context of another task, the task ends up with two context entities in its tags. It's indeterministic which one will be available through BrooklynTaskTags.getContextEntity. Add tests to cover the expected behaviour (which are failing atm) - the task should always have a single context entity, the one explicitly applied on the task before submitting.

neykov committed:
Fix resolving a sensor of an entity in the context of another entity.

The ExecutionContext will tag the task with the entity it belongs to. As it is tagged by the execution context no need to tag it explicitly.
The context of the task should be the entity we are resolving the sensor against, not the context of the current task.

neykov committed:
Don't allow multiple context entities in a task's tags.

The context entity should be set by internal code at a single place. Remove all places where we are setting the context additionally and add a warning at runtime if detected. To be changed to an exception after a release or two.

Commit:
Fix Javadoc references

aledsage committed:
BROOKLYN-356: threading for DSL task

Adds ImmediateSupplier, so can fetch Maybe<?> of DSL value immediately,
in current thread (rather than invoking in task with a timeout).

aledsage committed:
Rename JcloudsLocationExternalConfigYamlLiveTest

aledsage committed:
LocalManagementContextForTests supports properties map

aledsage committed:
Disable JavaWebAppWithDslYamlRebindIntegrationTest

Needs to be moved to brooklyn-library, or deleted

aledsage committed:
Reflections utils: improve use of generics

aledsage committed:
Convert jclouds yaml tests to unit tests

bostko committed:
Move instance.addPermissionsToLocation to a separate method
which can be tested easily with unit tests.

geomacy committed:
Add a stop latch to SoftwareProcess.

geomacy committed:
Add a stop latch to SoftwareProcess.

aledsage committed:
Rebind tests: avoid ClassCastException if multiple apps

aledsage committed:
Fix cancelling of BrooklynDslDeferredSupplier

DslAndRebindYamlTest was hanging - we were waiting for all tasks of that
entity to complete. However, the ValueResolver had kicked off a sub-task
which was then left running.

ahgittin committed:
fix Asserts missing import, tidy following review comments

ahgittin committed:
address Identifiers comment

neykov committed:
Break version range dependency on jackson-databind

com.maxmind.db:maxmind-db:jar:1.2.1 depends on jackson-databind [2.7.0,)
which causes excessive download of dependencies (all versions in the range,
including snapshots). We already include it explicitly anyway so can save on
the extra http requests. It's also forces maven to check frequently whether
new versions are available so it can use them.

drigodwin committed:
Enabled logging for specific exceptions

Commit:
DynamicGroup wraps entity filters in a same-app check

Commit:
BasicStartable is on fire if any of its children are on fire

And follow the same pattern as DynamicCluster for setting expected
state when starting and stopping.

Fixes https://issues.apache.org/jira/browse/BROOKLYN-372.

Commit:
AbstractApplication recovers from failure of children

Previously if a problem with child start up was fixed the application
would only recover if its start method was re-run.

aledsage committed:
This closes #389

neykov committed:
Closes #387

Convert jclouds yaml tests to unit tests

Commit:
Handle pre/post/launch errors in AbstractSoftwareProcessDriver restart

Previously if any phase errored (e.g. a post-launch comment exited with a
non-zero code) the entity would be stuck in the 'starting' state.

Fixes https://issues.apache.org/jira/browse/BROOKLYN-371

Commit:
Asserts.expectedFailureOfType requires at least one type

geomacy committed:
Review comment updates.

neykov committed:
Close #388

Add a stop latch to SoftwareProcess.

Adds the ability to delay stop of a software process until an appropriate
sensor is published on some entity that depends on that process.

neykov committed:
Closes #388

Add a stop latch to SoftwareProcess

Adds the ability to delay stop of a software process until an appropriate
sensor is published on some entity that depends on that process.

andreaturli committed:
add exist check before getting the blob

ahgittin committed:
This closes #394

ahgittin committed:
modify jclouds logging categories to fix certain OOME's

ahgittin committed:
This closes #395

Commit:
Add AbstractInvokeEffectorPolicy

Supports tracking the number of effector tasks that are ongoing and
publishing that as an "is busy" sensor on the entity the policy is
attached to.

neykov committed:
Closes #386

Add AbstractInvokeEffectorPolicy

Supports tracking the number of effector tasks that are ongoing and publishing that as an "is busy" sensor on the entity the policy is attached to. Tracking is enabled by providing a value for `isBusySensor` when configuring the policy.

neykov committed:
Closes #393

Lifecycle fixes

Fixes a handful of surprising behaviours in AbstractApplication, DynamicGroup and BasicStartable.

neykov committed:
Closes #391

Enable BasicExecutionManager logging for specific exceptions

A NullPointerException in ServiceReplacer was being swallowed in BasicExecutionManager, this makes exemptions for logging specific types of Exceptions.

neykov committed:
Closes #359

Utils tidy

tiny tidies to utils, including deprecation making `Maybe` a bit easier to work with in a couple places

neykov committed:
Closes #319

Clean up duplicated code in JcloudsLocationCustomizers

Unite the code from
https://github.com/apache/brooklyn-server/pull/276
and https://github.com/apache/brooklyn-server/pull/292
OPEN_INBOUND_PORTS_IN_SECURITY_GROUP_EFFECTOR tested after brooklyn restarts. All works fine.

neykov committed:
Closes #321

Added configuration for template substitution variable map

Adds a new `ConfigKey<Map<?,?>>` to allow setting further template variable substitutions in YAML.

neykov committed:
A test to check rebind when bundle prefixes are used to load classes in catalog items

neykov committed:
Fix master build

aledsage committed:
PR #390: incorporate review comments

aledsage committed:
PR #390: incorporate more review comments

aledsage committed:
More use of Tasks.resolving.immediately

Similar to BROOKLYN-356, use more efficient thread usage and avoid
premature timeouts if threads are starved.

neykov committed:
Closes #390

BROOKLYN-356: fix race in transformer by adding getImmediately

aledsage committed:
BROOKLYN-377: adds testConfigOnParentUsesConfigKeyDeclaredOnParent

aledsage committed:
Adds TestCaseYamlTest

neykov committed:
Closes #399

Adds TestCaseYamlTest

neykov committed:
Create a new management context for rebind test

The tests in RebindManagerExceptionHandlerTest were reusing the same management context on rebind - create a new management context instead to make sure we are rebinding against the on-disk persisted state.

neykov committed:
Stop persitence before rebind attempt

This is an attempt to fix test failures on the Apache Jenkins Windows slaves. The tests would frequrently fail with "The process cannot access the file because it is being used by another process" when trying to read the persisted state. Persistence in the old management context wasn't being stopped before trying the rebind so there's a chance that it would try to overwrite the files while the new management context rebinds, leading to file locking errors. Stopping persistence before rebind should solve the locking failures.

aledsage committed:
Add ConfigParametersYamlTest cases for top-level params

googlielmo committed:
Bump version of snakeyaml to 1.16

 - align to version pulled in by the `kubernetes-client` feature

aledsage committed:
BROOKLYN-380: Adds DynamicFabric.includeInitialChildren config

And adds tests for existing behaviour.

aledsage committed:
Move SanitizerTest to correct package

aledsage committed:
App defaults to requiring all children up

aledsage committed:
Fix Transformer (for when registered before entity managed)

The transformer can be added before the entity is managed (e.g. if
specified in the entity spec, as is often done via yaml). In that case,
the entity’s management context is still a “NonDeploymentManagementContext”.
The subscriptions are stored in a queue. When the entity is managed,
the subscriptions are replayed.

However, the queued subscriptions did not store the entity context.
Therefore when the transformer received its events, it didn’t have
an entity context so could not determine which entity it was talking
about.

This caused brooklyn-library’s EnrichersSlightlySimplerYamlTest.testWithAppEnricher
to fail.

aledsage committed:
This closes #402

aledsage committed:
This closes #400

aledsage committed:
Fix app.stop so doesn’t go on-fire briefly

Previously in app.stop, we set the serviceUp=false before setting
expectedState=stopping. This meant there was a race where we could
briefly infer the actual state as on_fire.

This caused failure of SoftwareProcessEntityLatchTest.testStopLatchBlocks:
the attributeWhenReady latch would abort because it received an
event saying that the app was on_fire, so the stop_latch completely
immediately.

The fix is simple: in app.stop, set the expectedSate=stopping before
setting serviceUp=false.

neykov committed:
Closes #404

Fix Transformer (for when registered before entity managed)

The transformer can be added before the entity is managed (e.g. if
specified in the entity spec, as is often done via yaml). In that case,
the entity’s management context is still a `NonDeploymentManagementContext`.
The subscriptions are stored in a queue. When the entity is managed,
the subscriptions are replayed.

However, the queued subscriptions did not store the entity context.
Therefore when the transformer received its events, it didn’t have
an entity context so could not determine which entity it was talking
about.

This caused brooklyn-library’s `EnrichersSlightlySimplerYamlTest.testWithAppEnricher`
to fail.

grkvlt committed:
Added test for transformer using parameters

aledsage committed:
Fix DSL config() to use parameter defaults

neykov committed:
Closes #405

Move SanitizerTest to correct package

ahgittin committed:
more conveniences for Strings

ahgittin committed:
let yaml tests look in the package where the class is defined
(so we can group things better than all in the root)

ahgittin committed:
new test showing changed behaviour of default config value resolution

as included in the release notes

ahgittin committed:
tidy warnings (deprecation, unused) for camp tests

aledsage committed:
Add ApplicationLifecycleStateStressTest

aledsage committed:
This closes #410

ahgittin committed:
comment, and license header, added to test case

aledsage committed:
Fix ServiceRestarterTest.testDoesNotSetOnFireOnFailure

The previous commit broke it, by changing FailingEntityImpl to
set itself as down/on-fire if restart() failed. Make that behaviour
configurable.

aledsage committed:
This closes #406

aledsage committed:
This closes #411

aledsage committed:
BROOKLYN-379: entity type includes OSGi prefix

neykov committed:
Fix integration tests in brooklyn-core

drigodwin committed:
Adds SoftRefLRUPolicyMSPerMB

Add -XX:SoftRefLRUPolicyMSPerMB=1 to testing params, see https://github.com/apache/brooklyn-dist/pull/58

aledsage committed:
Fix osgi resolveClass to strip bundle prefix

Without this, RebindOsgiTest.testUsesCatalogBundleVersion fails

aledsage commented on core/src/main/java/org/apache/brooklyn/core/mgmt/ha/OsgiManager.java:
```diff
@@ -189,10 +193,11 @@ public static boolean isBundleNameEqualOrAbsent(CatalogBundle bundle, Bundle b)
                     Bundle b = bundle.get();
-                    Class<T> clazz;
+                    Optional<String> strippedType = osgiClassPrefixer.stripMatchingPrefix(b, type);
+                    String typeToLoad = strippedType.isPresent() ? strippedType.get() : type;
```
Not sure if this is the right place for the fix! Arguably if it has the prefix, we should only look in one bundle (or multiple versions of the bundle), rather than iterating over all bundles.

neykov replied:
If a class has bundle prefix obviously it's not relying on the libraries functionality to load it. Still could be useful to have this to avoid version mismatch (half of the blueprint is coming from libraries section the other half is using prefixes because of copy&paste).
I'd structure it a bit differently - if there's a prefix short-circuit and don't try to load it from non-matching bundles, but can't get away from iterating over them. And it's not **all**, just those in the libraries section so it's fine.
The prefix could include version as well - would be nice to support it here as well (by skipping non-matching versions) to avoid introducing subtle inconsistencies in the prefix behaviour.

neykov replied:
This makes more sense now that I saw the commit comment. This is called by persistence, not only by yaml parsing code.
Another approach would be to skip writing the prefix if it the class comes from bundles in the memento's catalog item.

aledsage commented:
I'm still running `mvn clean install` to check the rest of the tests. Will report back later this evening, when that's finished.

Commit:
Typos

Commit:
XmlSerialiserTest logs at debug

\u001b in testIllegalXmlCharacter breaks tmux when it's printed

neykov committed:
Closes #413

Adds SoftRefLRUPolicyMSPerMB

Add -XX:SoftRefLRUPolicyMSPerMB=1 to testing params, see https://github.com/apache/brooklyn-dist/pull/58

tbouron committed:
TestHttpCall entity can now perform requests other than GET, setup headers and body

aledsage committed:
DynamicFabric.includeInitialChildren: address review comments

tbouron committed:
Add tests + address PR comments

neykov committed:
Convert integration tests to unit tests and fix timing related failures

neykov committed:
Use ssh cli in batch mode - don't ask for passwords

aledsage committed:
Code tidy for tests

aledsage committed:
Add more DSL tests

aledsage committed:
Adds TestHttpCallYamlTest

aledsage committed:
Fix timeout in ValueResolver.immediate(false)

If not immediate, should not be setting timeout to 500ms

aledsage committed:
Deprecate DslComponent(String)

aledsage committed:
BROOKLYN-381: nested DSL support in $brooklyn:entity()

For example, $brooklyn:entity(attributeWhenReady(“targetId”))

Also tests this in the context of a TestHttpCall, in TestHttpCallTest

Commit:
Delete duplicate dependency on jackson-datatype-guava

aledsage committed:
TestCaseYamlTest: increase assertion timeout

Commit:
Fix Debian 7 live test

ami-5586a43c is not unavailable.

aledsage committed:
This closes #416

aledsage committed:
This closes #403

aledsage committed:
Adds ConfigYamlTest.testConfigAtTopLevel

aledsage committed:
Fix SpecParameterUnwrappingTest

The number of default config keys on app had changed, breaking the tests.

ahgittin committed:
testing map config dot syntax, isolating known lookup failure

ahgittin committed:
REST API supports client requiring a CSRF header, and

requesting such a header, and if required POST requests fail if it wasn't supplied

ahgittin committed:
resolve values using the key declared on an entity etc

ahgittin committed:
display info on started apps in the log / stdout

handy with the docker one-line deployer

Commit:
SshCommandEffector resolves shell.env against the target entity

Rather than always resolving it against the entity the effector is
attached to. This also fixes the resolution of the execution directory.

neykov committed:
Closes #419

Fix key lookup

fixes subtle bug which prevents some evaluation strategies for map subkeys

aledsage committed:
Add CatalogPredicates toString methods

aledsage committed:
Improve CatalogEntitySummary.toString

aledsage committed:
Add pre/post-job to PerformanceMeasurer

aledsage committed:
Add catalog performance tests

aledsage committed:
CatalogResource: minor tidy

aledsage committed:
BROOKLYN-382: Add BasicBrooklynCatalog.specCache

Cache the Specs we generate (by parsing the yaml), rather than
re-parsing the YAML every time createSpec(item) is called.

neykov committed:
Closes #421

Show info after application start

handy with the docker one-line deployer

neykov committed:
Upgrade testng to latest version

Now that we don't support Java 6 there's no restriction on upgrading testng.
The new version could improve the execution order by not interleaving tests from different classes. But seems that's not the case when using dependsOn*

neykov commented on core/src/main/java/org/apache/brooklyn/core/mgmt/persist/OsgiClassPrefixer.java:
```diff
@@ -0,0 +1,59 @@
+    public Optional<String> stripMatchingPrefix(Bundle bundle, String type) {
+        String symbolicName = bundle.getSymbolicName();
+        if (symbolicName != null && type.startsWith(symbolicName + ":")) {
+            return Optional.of(type.substring((symbolicName + ":").length()));
```
The prefix could include version identifier as well.

neykov commented on core/src/main/java/org/apache/brooklyn/core/mgmt/persist/OsgiClassPrefixer.java:
```diff
@@ -0,0 +1,59 @@
+    
+    public Optional<String> getPrefix(Class<?> type) {
+        Optional<Bundle> bundle  = Osgis.getBundleOf(type);
+        if (bundle.isPresent() && !whiteListRetriever.isBundleWhiteListed(bundle.get())) {
```
Don't like how the white list is proliferating throughout the code base. It's supposed to go the other way :).
What do we gain by stripping the prefix for white listed bundles? I think not much because if classes move bundles then their package will change anyway. This can be handled by the renaming list.

Removing this will simplify the code and we don't lose much if anything.

neykov commented:
Would be interested to hear your thoughts on the above comments.

aledsage committed:
BROOKLYN-383: fix json serialisation of tasks

* Also do the same for policies, enrichers and feeds.
* Adds AbstractEntityAdjunct.getEntity() to make that easier.
* Adds TestEntity.sleepEffector

aledsage committed:
Test XML persistence of task

aledsage committed:
This closes #385

aledsage committed:
BROOKLYN-304: avoid json err serializing ClassLoader

aledsage committed:
This closes #422

aledsage committed:
This closes #418

aledsage committed:
This closes #414

aledsage committed:
This closes #401

aledsage committed:
Improve robustness of MemoryUsageTrackerTest

neykov committed:
brooklyn-locations-jclouds integration test fixes

neykov committed:
Closes #426

Improve robustness of MemoryUsageTrackerTest

Motivated by discussion in https://github.com/apache/brooklyn-server/pull/412 with @neykov (cc @ahgittin).

neykov committed:
more integration test fixes

brooklyn-software-base, brooklyn-camp, brooklyn-rest-resources

googlielmo committed:
Use servicemix bundle for okio

 - servicemix okio version 1.2.0_1
 - still wrap: okhttp 2.2.0 but with explicit dependency on okio 1.2.x bundle

aledsage committed:
Fix SpecParameterParsingTest

(number of default config keys in app had changed)

neykov committed:
Closes #398

BROOKLYN-377: adds testConfigOnParentUsesConfigKeyDeclaredOnParent

This PR contains a failing test, which demonstrates the underlying problem in https://issues.apache.org/jira/browse/BROOKLYN-377.

I believe that https://github.com/apache/brooklyn-server/pull/385 will fix this. We should rebase + retest at that point. In particular, if I change `ValueResolver` to delete `.tagIfNotNull(BrooklynTaskTags.getTargetOrContextEntityTag(Tasks.current()))` then the test passes.

@neykov @drigodwin you'll both likely be interested in this.

aledsage committed:
This closes #427

neykov committed:
Make sure entity doesn't equal null

aledsage committed:
This closes #412

aledsage committed:
This closes #417

ahgittin committed:
switch CSRF to use cookies for tokens

now supports AngularJS semantics.
also now it doesn't needlessly create sessions.

ahgittin committed:
set session/cookie on some server requests so client gets it early

helps establish csrf protection. done in /server/user and /server/up/extended,
the two main places which an interactive app will hit early.

ahgittin committed:
logout rest code tidy - behaves nicer if no user

previously `curl /v1/logout` would throw 500 server error

ahgittin committed:
tidy - warnings / unused imports in rest

neykov committed:
Initialize REST API security correctly for integration tests

Also add checks for it in the tests.

Tests had different behaviour depending on whether brooklyn-ui existed along brooklyn-server on disk. When brooklyn-ui was there tests would find it and use the web app code from there, including the web.xml which restricts requests to only authorized users. When brooklyn-ui is missing though a default web app without web.xml was being created. This would result in non-authorized requests succeeding.  In this case requests including the authorization header would be accepted and rejected if the password is invalid. But in order to include the authorization header the server must first respond with a 401 which didn't happen. Moving web-security.xml to rest-server allows us to force request authentication for tests even if there's no web.xml in the web app.
Should be fine moving web-security.xml to rest-server (which is only used in classic) because it's only inserted by classic related code. Karaf Brooklyn doesn't have control over the web apps.

neykov committed:
Don't cache the delegatinProvider in a static field as that's causing test failures (when reloading properties with a different security provider). DelegatingSecurityProvider caches the delegate in mgmt so re-creating it for the same mgmt is cheap.

neykov committed:
Non-HOT HA modes don't allow app-related requests in any more.

aledsage committed:
SoftwareProcess integration tests: use INSTALL_UNIQUE_LABEL

Fixes VanillaSoftwareProcessStreamsIntegrationTest.testGetsStreams,
for if run multiple times on the same machine.

aledsage committed:
DslComponent: fix typo in comment

aledsage committed:
This closes #432

neykov committed:
Clean up BrooklynJacksonSerializerIntegrationTest and mark test as Broken

tbouron committed:
Fix dependency URI path for downstream projects + Karaf feature verification

neykov committed:
Don't lose the security provider configured in tests on reload

aledsage committed:
This closes #433

aledsage committed:
BROOKLYN-382: use catalog.peekSpec()

* Deprecates catalog.createSpec()
* Adds catalog.peekSpec(), to return a spec object that is not intended
  for creating apps, but instead for just inspecting additional info
  of the item.
* Tidies up the BasicBrooklynCatalog.specCache
  (e.g. invalidate it all if anything changes in the catalog).
* Simplifies the generics of catalog.peekSpec, so that callers don’t
  do horrible casting to remove generics!

neykov committed:
Minor integration test fixes

* Request the correct "skipSecurity" mode
* Platform independent ssh result test
* Upgrade surefire to latest available

aledsage committed:
Move INTERNAL_PROVISIONING_TASK_STATE to AttributesInternal

So that it can subsequently be used to fix BROOKLYN-325.

aledsage committed:
BROOKLYN-325: alert if provisioning/termination aborted on rebind

aledsage committed:
This closes #431

aledsage committed:
EntityExecutionManagerTest: tidy

neykov committed:
Set default JVM memory settings for the build

aledsage committed:
BROOKLYN-386: Avoid NPE in JcloudsSshMachineLocation.getOptionalNode

aledsage committed:
Asserts: use DEFAULT_LONG_TIMEOUT in eventually()

And update javadoc to say why. Also makes the timeout configurable
via `-Dbrooklyn.test.defaultTimeout=1s`.

aledsage committed:
BROOKLYN-386: avoid locationAdded events on rebind

aledsage committed:
BROOKLYN-386: adds CreateUserPolicyRebindTest

Commit:
Add Suppliers.incrementingSupplier

neykov committed:
Exclude .mvn/jvm.config from rat checks

Also merge pluginManagement in the project configuration for rat-plugin

neykov committed:
Closes #437

Asserts: use DEFAULT_LONG_TIMEOUT in eventually()

And update javadoc to say why. Also makes the timeout configurable
via `-Dbrooklyn.test.defaultTimeout=1s`.

neykov committed:
Closes #407

App defaults to requiring all children up

neykov committed:
Closes #439

Add Suppliers.incrementingSupplier

A simple useful class. I have a pull request coming for brooklyn-library that makes proper use of this.

neykov committed:
Closes #438

Fix BROOKLYN-386: avoid locationAdded events on rebind

Commit:
This closes #436

Set default JVM memory settings for the build

Commit:
This closes #424

BROOKLYN-383: fix json serialisation of tasks

aledsage committed:
BROOKLYN-325: address review comments

aledsage committed:
Performance tests: minor tidy up

neykov committed:
Closes #423

BROOKLYN-382: get-catalog performance improvements

The `specCache` speeds up the REST api calls to `/catalog/entities` considerably.

For example, with 10 small yaml entities in the catalog, before the cache it takes about 170ms per call (and proportionately more as we increase the number of catalog items or their size):

```
test=PerformanceTestDescriptor{summary=CatalogResourcePerformanceTest.testListAllEntities, iterations=100, warmupIterations=10, job=org.apache.brooklyn.rest.resources.CatalogResourcePerformanceTest$8@246f150f, minAcceptablePerSecond=1.0}; result=PerformanceTestResult{duration=16s 981ms, warmup=3s 274ms, iterations=100, warmupIterations=10, ratePerSecond=5.888934691714269, histogram={134.2ms-268.4ms: 100}, cpuTotalFraction=1.3575868461219567, cpuSampleFractions=[1.2573016983016982, 1.9225778443113772, 1.4971834496510468, 1.5281373134328358, 1.1606623505976095, 1.2851962151394423, 1.2523649052841475, 1.2795744255744255, 1.211623880597015, 1.1832784431137724, 1.162554, 1.3302412761714855, 1.1046873126873127, 1.4044291417165669, 1.0696699900299103, 1.846793, 1.0917181274900398], minAcceptablePerSecond=1.0}
```

With the cache, it takes 24ms per call:
```
test=PerformanceTestDescriptor{summary=CatalogResourcePerformanceTest.testListAllEntities, iterations=100, warmupIterations=10, job=org.apache.brooklyn.rest.resources.CatalogResourcePerformanceTest$8@345a656a, minAcceptablePerSecond=1.0}; result=PerformanceTestResult{duration=2s 417ms, warmup=1s 393ms, iterations=100, warmupIterations=10, ratePerSecond=41.373603640877114, histogram={8.389ms-16.78ms: 20, 16.78ms-33.55ms: 74, 33.55ms-67.11ms: 4, 67.11ms-134.2ms: 1, 134.2ms-268.4ms: 1}, cpuTotalFraction=2.552031110340871, cpuSampleFractions=[2.367437, 1.722136], minAcceptablePerSecond=1.0}
```

ahgittin committed:
This closes #430

aledsage committed:
This closes #435

aledsage committed:
This closes #434

aledsage commented:
I'm going to move this into its own apache/brooklyn-server PR, to replace https://github.com/apache/brooklyn-server/pull/396. That way the discussion will be more visible to the brooklyn community.

I'll replay your comments there.
